### PR TITLE
Issue 609

### PR DESCRIPTION
**Pull request by HoshinoTented**
This PR:

- delete `Pat#toExpr` ~~(unused for now, but it is safe to delete)~~
- insert implicit argument when `PatToTerm`
- improve replacement when `PatTyck` (see `ExprTycker#lets`)

### CODE REVIEW

File: base/src/main/java/org/aya/tyck/pat/PatTycker.java [hunk 1]
```diff
@@ -374,9 +399,18 @@ private record PatData(
     tracing(TreeBuilder::reduce);
     addPatSubst(data.param.ref(), res, pat.sourcePos());
     data.results.append(res);
+
+    // update signature for current parameter,
+    // because it probably changed when we call `data.sig.inst`
+    exprTycker.localCtx.put(data.param().ref(), type);
```
Comment (ice1000) [inline base/src/main/java/org/aya/tyck/pat/PatTycker.java:405]: I suspect that the extra bindings come from here.

Comment (HoshinoTented) [reply base/src/main/java/org/aya/tyck/pat/PatTycker.java:405]: of course

**Comment by ice1000** [comment]
https://github.com/aya-prover/aya-dev/blob/c7c57e4ee13577999163d0ef06864f829d7ae7cf/.github/workflows/commit-check.sh#L12

**Comment by ice1000** [comment]
I suggest `pat: improve code` but it's up to you

**Comment by HoshinoTented** [comment]
> I suggest `pat: improve code` but it's up to you

solved

File: base/src/main/java/org/aya/tyck/pat/PatTycker.java (hunk 1 above)
Comment (HoshinoTented) [reply base/src/main/java/org/aya/tyck/pat/PatTycker.java:405]: Don't resolve, this is a bug.

Comment (ice1000) [reply base/src/main/java/org/aya/tyck/pat/PatTycker.java:399]: I'm taking care of it

**Comment by HoshinoTented** [comment]
We have to display all var in `ExprTycker#lets` for users when context information is requested.

**Comment by HoshinoTented** [comment]
We didn't inline any `Term.MetaPat` in the types of Pat, I don't know if it makes any problem.

**Comment by ice1000** [comment]
@HoshinoTented 应该 inline 的

**Comment by HoshinoTented** [comment]
I found that this code won't compile

```
open data Foo (A : Type)
| B => foo A

def bar {A : Type} (x : Foo A) : A
| foo y => y
```

because we didn't subst the A to B. It seems like we need another `checkRhs` for Ctor, but this should be completed in another PR.

**Comment by HoshinoTented** [comment]
More extreme tests are NEEDED!

**Comment by ice1000** [comment]
codecov r+

**Pull request merged**